metricflow/model/validations/reserved_keywords.py

```diff
@@ -0,0 +1,162 @@
+from typing import List
+from metricflow.dataflow.sql_table import SqlTable
+from metricflow.instances import DataSourceElementReference
+
+
+from metricflow.model.objects.data_source import DataSource
+from metricflow.model.objects.user_configured_model import UserConfiguredModel
+from metricflow.model.validations.validator_helpers import (
+    DataSourceContext,
+    DataSourceElementContext,
+    DataSourceElementType,
+    FileContext,
+    ModelValidationRule,
+    ValidationError,
+    ValidationIssueType,
+    validate_safely,
+)
+
+# A non-exaustive tuple of reserved keywords
+# This list was created by running an intersection of keywords for redshift,
+# postgres, bigquery, and snowflake
+RESERVED_KEYWORDS = (
+    "AND",
+    "AS",
+    "CREATE",
+    "DISTINCT",
+    "FOR",
+    "FROM",
+    "FULL",
+    "HAVING",
+    "IN",
+    "INNER",
+    "INTO",
+    "IS",
+    "JOIN",
+    "LEFT",
+    "LIKE",
+    "NATURAL",
+    "NOT",
+    "NULL",
+    "ON",
+    "OR",
+    "RIGHT",
+    "SELECT",
+    "UNION",
+    "USING",
+    "WHERE",
+    "WITH",
+)
+
+
+class ReservedKeywordsRule(ModelValidationRule):
+    """Check that any element that ends up being selected by name (instead of expr) isn't a commonly reserved keyword.
+
+    Note: This rule DOES NOT catch all keywords. That is because keywords are
+    engine specific, and semantic validations are not engine specific. I.e. if
+    you change your underlying data warehouse engine, semantic validations
+    should still pass, but your data warehouse validations might fail. However,
+    data warehouse validations are slow in comparison to semantic validation
+    rules. Thus this rule is intended to catch words that are reserved keywords
+    in all supported engines and to fail fast. I.e. `USER` is a reserved keyword
```

Comment:
e.g. means "for example" and i.e. means "that is"

```suggestion
    in all supported engines and to fail fast. E.g., `USER` is a reserved keyword
```

My inner pedant is satisfied. Thanks for your patience.